### PR TITLE
[WMS][12.0] Add stock_move_location_dest_constraint_base - alpha version

### PR DESCRIPTION
**Pull request by grindtildeath**
This module provides a base mechanism to constrain the selection of destination
location on stock move lines.
It also adds a checkbox 'Bypass constraints' on stock location to deactivate
the checks implemented by installed modules.

Related to https://github.com/OCA/stock-logistics-warehouse/issues/691

### CODE REVIEW

File: stock_move_location_dest_constraint_base/models/stock_location.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        self.ensure_one()
+        return True
+
+    @api.model
+    def _set_bypass_on_existing_locations(self):
+        """Set bypass_constrains on all the existing locations"""
+        existing_locations = self.search([])
+        existing_locations.write({'bypass_constraints': True})
+
+    def filtered(self, func):
```
Comment (grindtildeath) [inline stock_move_location_dest_constraint_base/models/stock_location.py]: :warning: This must probably be moved to model `stock.fixed.putaway.strat`

**Comment by jgrandguillaume** [comment]
@grindtildeath I have the following error while trying to receive products.

To replicate this, I have a put away in input location (with bybass constraints set) to put it away in input /WAE for all product category.

`
Traceback (most recent call last):
  File "/odoo/src/odoo/http.py", line 654, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/odoo/src/odoo/http.py", line 312, in _handle_exception
    raise pycompat.reraise(type(exception), exception, sys.exc_info()[2])
  File "/odoo/src/odoo/tools/pycompat.py", line 87, in reraise
    raise value
  File "/odoo/src/odoo/http.py", line 696, in dispatch
    result = self._call_function(**self.params)
  File "/odoo/src/odoo/http.py", line 344, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/odoo/src/odoo/service/model.py", line 97, in wrapper
    return f(dbname, *args, **kwargs)
  File "/odoo/src/odoo/http.py", line 337, in checked_call
    result = self.endpoint(*a, **kw)
  File "/odoo/src/odoo/http.py", line 939, in __call__
    return self.method(*args, **kw)
  File "/odoo/src/odoo/http.py", line 517, in response_wrap
    response = f(*args, **kw)
  File "/odoo/src/addons/web/controllers/main.py", line 966, in call_button
    action = self._call_kw(model, method, args, {})
  File "/odoo/src/addons/web/controllers/main.py", line 954, in _call_kw
    return call_kw(request.env[model], method, args, kwargs)
  File "/odoo/src/odoo/api.py", line 749, in call_kw
    return _call_kw_multi(method, model, args, kwargs)
  File "/odoo/src/odoo/api.py", line 736, in _call_kw_multi
    result = method(recs, *args, **kwargs)
  File "/odoo/src/addons/purchase/models/purchase.py", line 331, in button_confirm
    order.button_approve()
  File "/odoo/src/addons/purchase_stock/models/purchase.py", line 87, in button_approve
    self._create_picking()
  File "/odoo/src/addons/purchase_stock/models/purchase.py", line 214, in _create_picking
    moves._action_assign()
  File "/odoo/external-src/stock-logistics-warehouse/stock_picking_zone/models/stock_move.py", line 11, in _action_assign
    super()._action_assign()
  File "/odoo/src/addons/stock/models/stock_move.py", line 914, in _action_assign
    self.env['stock.move.line'].create(move._prepare_move_line_vals(quantity=missing_reserved_quantity))
  File "/odoo/src/addons/stock/models/stock_move.py", line 803, in _prepare_move_line_vals
    location_dest_id = self.location_dest_id.get_putaway_strategy(self.product_id).id or self.location_dest_id.id
  File "/odoo/src/addons/stock/models/stock_location.py", line 132, in get_putaway_strategy
    putaway_location = current_location.putaway_strategy_id.putaway_apply(product)
  File "/odoo/src/addons/stock/models/product_strategy.py", line 29, in putaway_apply
    put_away = self._get_putaway_rule(product)
  File "/odoo/external-src/stock-logistics-warehouse/stock_move_location_dest_constraint_base/models/product_strategy.py", line 15, in _get_putaway_rule
    return super()._get_putaway_rule(product)
  File "/odoo/src/addons/stock/models/product_strategy.py", line 42, in _get_putaway_rule
    put_away = self.fixed_location_ids.filtered(lambda x: x.category_id == categ)
  File "/odoo/external-src/stock-logistics-warehouse/stock_move_location_dest_constraint_base/models/product_strategy.py", line 31, in filtered
    put.location_id.check_move_dest_constraint(product=product)
AttributeError: 'stock.fixed.putaway.strat' object has no attribute 'location_id'
`

File: stock_move_location_dest_constraint_base/models/product_strategy.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    def filtered(self, func):
+        """Filter putaway strats according to installed constraints"""
+        putaway_strats = super().filtered(func)
+        if self.env.context.get('_filter_on_constraints'):
+            product_id = self.env.context.get('_constraint_product')
+            product = self.env['product.product'].browse(product_id)
+            filtered_putaways = self.browse()
+            for put in putaway_strats:
+                try:
+                    put.location_id.check_move_dest_constraint(product=product)
```
Comment (grindtildeath) [inline stock_move_location_dest_constraint_base/models/product_strategy.py]:
```suggestion
                    put.fixed_location_id.check_move_dest_constraint(product=product)
```

Comment (jgrandguillaume) [reply stock_move_location_dest_constraint_base/models/product_strategy.py]: I confirm this fix the problem

**Review by jgrandguillaume — APPROVED** [review]
Let' merge this now. We'll bring other improvemnts later on for perfe reasons

**Pull request merged**